File: 1371-find-the-longest-substring-containing-vowels-in-even-counts/1371-find-the-longest-substring-containing-vowels-in-even-counts.ts
function findTheLongestSubstring(s: string): number {
  const v: Readonly<Record<string, number>> = { 'a': 1, 'e': 2, 'i': 4, 'o': 8, 'u': 16 },
        m: Record<number, number> = { 0: -1 };
  let mask = 0,
      ans = 0;
  for (let i = 0; i < s.length; i++) {
    if (s[i] in v) mask ^= v[s[i]];
    if (mask in m) {
      if (ans < i - m[mask]) ans = i - m[mask];
    } else {
      m[mask] = i;
    }
  }
  return ans;
};

File: 1371-find-the-longest-substring-containing-vowels-in-even-counts/1371-find-the-longest-substring-containing-vowels-in-even-counts.go
func findTheLongestSubstring(s string) int {
  v := map[rune]int{ rune('a'): 1, rune('e'): 2, rune('i'): 4, rune('o'): 8, rune('u'): 16 }
  m := map[int]int{ 0: -1 }
  mask := 0
  ans := 0
  for i, r := range s {
    if bit, ok := v[r]; ok {
      mask ^= bit
    }
    if pos, ok := m[mask]; ok {
      if diff := i - pos; ans < diff {
        ans = diff
      }
    } else {
      m[mask] = i
    }
  }
  return ans
}